### ISSUE TITLE
Resetting a form does not fire `onChange`

### ISSUE THREAD

**Issue by jmeas** [davidkpiano/react-redux-form#858]
### The Problem

Calling `resetForm` does not call `onChange` on controls.

This may be a problem, or maybe not. The behavior was unexpected to me, but it may just be that I'm still learning the contract of `onChange`.

### Steps to Reproduce

1. Put a control in a form
2. Reset the form
3. Observe that `onChange` isn't called when the model is reset

### Expected Behavior

That `onChange` would be called with the value after the reset

### Actual Behavior

`onChange` is not called is not called with the value after the reset

### Reproducible Code Example

[View here](https://codepen.io/anon/pen/JJyvXX)

Change the "First Name" field to see the `onChange` event called.
Then, click the button to reset the form. Observe that `onChange` is not called, even though the input value _does_ change.

@davidkpiano – I understand that this may be the intended behavior, which is alright. If that's the case, what would you suggest I do instead?

The use case is that I have some custom controls that need to manage internal state based on the value of the model. They're kept in sync with `onChange`. I was thinking I could use this reliably to keep the state in sync, but perhaps I should wire them up with `connect`, and attempt to use `mapStateToProps` to get the value? Or, do you have another suggestion?

As always – thank you for taking a look at this! Working with RRF has been great 👌

**Comment by davidkpiano** [comment]
It is the intended behavior, because the semantics of changing a model and resetting it are a little different, in that in most use-cases, you want to handle when the user _modifies_ a model, and it's sort of a gray area as to whether or not resetting counts as modifying a model.

What do others think? Should the `change` action fire when a model is reset?

**Comment by jmeas** [comment]
Makes sense, @davidkpiano . Are there any hooks for when a value changes due to resetting? If not, do others think that's a problem, independent of whether the notification happens through `onChange` or not?

This causes an issue to me because the thing that resets the form is usually higher up than the controls themselves. There's no simple way for me to "push" action that reset down into the component, other than by accessing the value model value from the store. I haven't put _too_ much thought into it, but it seems like a challenging project to get the value of a custom control that accepts the `".model"` syntax from _outside_ of the component, say, in a `mapStateToProps` for the component that wraps it. My guess (I haven't dug into the code) is that the form puts some stuff on context, which the control then accesses. So I suppose I could access the context directly in my app to find the model, but that doesn't seem as elegant.

Anyway, I'm not pushing for `onChange` to change, or anything. I'm open to any solution. If there's a recommended way to do this with the current API, then I'm definitely :+1: to that.

**Comment by jmeas** [comment]
The above comment may have come across as rambling (I admit I was thinking through the issue as I was typing). Here's a simple challenge that illustrates the problem:

- Build a component that wraps `Control.custom`
- Within that component's `render` function, `console.log()` the current value of that Control's model

One way to do this is to use the `onChange` hook to set some state on the component equal to the new value passed through `onChange`. Then, you would log out that value in `render`. Because resets don't fire `onChange` events, the component's state drifts from the control's value.

Here's a little snippet to show what I mean:

```jsx
class CustomSearch extends Component {
  render() {
    // This may just be `".myModel"`, so using it to get the model's value
    // in `mapStateToProps` isn't trivial
    const { model } = this.props;
    const { selected } = this.state;
    
    // When the form is reset, this console.log will be incorrect
    console.log('This control\'s current value is', selected);

    return (
      <Control.custom
        model={model}
        component={SomeComponent}
        value={value}
        mapProps={{
          onChange: props => selected => {
            this.setState({ selected });
            props.onChange(selected);
          }
        }}/>
    );
  }
}

```

---

My current solution is to no longer use the `reset` action creator.

What I'm instead doing is manually changing each value in all of my forms to their initial value to mimic the reset. It's much more verbose, and requires updating the reset code anytime that the form updates, but it's one way to get around the problem for now.

**Comment by davidkpiano** [comment]
Here is a related issue: https://github.com/facebook/react/issues/1152

**Comment by jmeas** [comment]
Makes sense @davidkpiano . Thanks for the cross-reference!

I haven't tried this yet, but I'm pretty confident that the solution is to always have `Control.Custom` as the "highest" component in a series of wrapped components. This way, it's connected for you, and passes the value of the model down to all of the children.

I'm going to close out this issue for now. Thanks for all of the help, @davidkpiano !

**Comment by jmeas** [comment]
Alright, so, update – I finally had a chance to refactor this for one of my inputs, and it ended up working. Here's a thorough description of the problem and the solution:

First, I have a Dropdown component (similar to React Select). It sometimes displays text determined by the model's current value, such as when you've selected an option. But since it's a typeahead-style component, sometimes it instead displays whatever you've been typing. This is stored in the component's internal state.

There are two ways to handle this: one is to write a conditional to determine when to render state, or when to render the model's value. The other way is to _always_ render the component's state, and then write code that keeps the state in sync with the model's value.

When I opened this issue, I was attempting to go with the latter approach. But I am now using the former approach.

The tl;dr of my solution is:

1. I set `focused` state in `onBlur`/`onFocus` events
2. I set whatever the user types as `value` state on the component, as well as an associated `selectedKey` value on state
2. In the `onBlur` callback, check to see if the state's `selectedKey` value matches the model's value. if not, I reset the state's `value` to match the model's `value`. This is so if the user goes into an input that already has an option selected, and they, say, type "asdf," then navigate away without selecting a new option, it goes back to what was selected before, rather than continuing to display "asdf".
3. within render, if the input is not focused, then I display the model's value. if the input is focused, then I display the value from the state.

Because the component defaults to using the model's actual value unless the user is typing, then you get free synchronization with the model, which, well, makes a lot of sense in hindsight.

There are two takeaways from this:

1. Do not wrap `Control` or `Control.Custom` with additional functionality. Keep those things pretty simple, and add additional functionality in the component that they wrap

2. Do not attempt to keep state in sync with a model's value. If you want to use a model's value, **use it directly**. If there are situations where you want to use some other value, then it is a much simpler problem to detect those situations.

If anyone else runs into a similar problem, or has further questions, just leave a comment here and I'd be happy to help out.

Thanks again for all of the help @davidkpiano !
